File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    conversation = Conversation.find(params[:conversation_id])
    stream_for conversation
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    conversation = Conversation.find(data["conversation_id"])
    message = conversation.messages.create!(content: data["content"], user: current_user)
    broadcast_message(conversation, message)
  end

  private

  def broadcast_message(conversation, message)
    ChatChannel.broadcast_to(conversation, {
      content: message.content,
      current_user_id: current_user.id,
      user: {
        id: message.user.id,
        username: message.user.username
      }
    })
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "./consumer"

document.addEventListener('turbolinks:load', () => {
  const conversationElement = document.getElementById('conversation-id')
    console.log('conversationElement: ', conversationElement)

  if (!conversationElement) return;

  const conversationId = conversationElement.getAttribute('data-conversation-id')

  if (conversationId) {
    const chatChannel = consumer.subscriptions.create(
        { channel: "ChatChannel", conversation_id: conversationId },
        {
          received(data) {
            console.log('Received data:', data)
            const messagesContainer = document.getElementById('messages')
              const messageHTML = `
            <div class="chat ${data.current_user === data.message.user_id ? 'chat-start' : 'chat-end'}"  data-user-id="${data.message_user}">
              <div class="chat-bubble"><p>${data.message.content}</p></div>
            </div>
          `;
            messagesContainer.insertAdjacentHTML('beforeend', messageHTML)
          },
          speak: function(content) {
            return this.perform('receive', { conversation_id: conversationId, content: content })
          }
        }
    )

    const messageForm = document.getElementById('new_message')
    messageForm.addEventListener('submit', function(event) {
      event.preventDefault()
      const messageContent = event.target.querySelector('#message_content').value
      chatChannel.speak(messageContent)
      event.target.reset()
    })
  }
})
